Refractor props for blending (#81)
brown-ccv/react-volume-viewer: README.md

npm install react-volume-viewer
```

## Props

> The only required props are the model's path, colorMap, and it's minimum and maximum data points. The model's path should be imported into the project and passed in from there - see the [example project](#example).

Reviewer: ```suggestion
The only required props are the model's path, name, and its minimum and maximum data points. The model's path should be imported into the project and passed in from there - see the [example project](#example).
```

> The only required props are the model's path, colorMap, and it's minimum and maximum data points. The model's path should be imported into the project and passed in from there - see the [example](#example) project.

Reviewer: ```suggestion
The only required props are the model's path, colorMap, and it's minimum and maximum data points. The model's path should be imported into the project and passed in from there - see the [example project](#example).
```

Reviewer: is `useColorMap` is false, is `colorMap` still required?

> CSS styling for the height must be provided and a custom width can be provided as well. The styling can be passed in through classes or inline styles.

Reviewer: which classes?

Author: Any class the user defines on their end. I reworded it to make that a little more obvious but I'm not sure the verbiage is correct

```jsx

VolumeViewer.propTypes = {
  /** Blending algorithm to apply between the models
   *    Blending.None: Don't apply blending
   *    Blending.Add: Apply additive blending
   */

>   blending: PropTypes.instanceOf(Blending),

Reviewer: now that `Blending` is just an object, does this check still work?

Author: Oops I'll update that in the README, it's correct in `VolumeViewer.jsx`

  /** Whether or not the controls can be seen */
  controlsVisible: PropTypes.bool,

>   /** Array of models loaded into aframe REQUIRED

Reviewer: maybe call out which of these fields is required?

Author: Great idea, will do!

  /** Array of models loaded into aframe (REQUIRED)
   *    colorMap: Object containing the path to the current color image applied to the model.
   *      name: Common name of the color map (REQUIRED)
   *      path: Path to the color map source image (REQUIRED)

>    *    colorMaps: Array of possible color maps for the model

Reviewer: indenting looks funny in this section

   *      Each colorMap must have a unique name in colorMaps
   *    description: Short description of the model
   *    enabled: Flag to display the model
   *    intensity: Multiplication factor for voxels intensity
   *    name: Unique name given to the model
   *    path: Path to the model
   *    range: Minimum and maximum values of the model's dataset
   *    transferFunction: The transfer function applied to the color map

>    *      transferFunction is an array of {x: [val], y: [val]} coordinates.

Reviewer: are the x and y values arrays? or is the `[val]` meant to be placeholder syntax? if placeholder syntax I'd maybe use `<val>` instead

Author: Yes they're placeholders, will use `<val>`

   *      Each coordinate in transferFunction must be between (0, 0) and (1,1)
   *    initTransferFunction: Initial value of transferFunction passed to the component
   *    useTransferFunction: Flag to apply a transfer function to the model
   *    useColorMap: Flag to apply a color map to the model
   */
  models: MODEL,

>   /** Position of the dataset in the scene as a "[x] [y] [z]" string

Reviewer: same comment on `[ ]` for placeholders.  That notation is often used to denote either that the field is optional, or it is an array, and I don't _think_ either is the case here

Author: The prop as a whole is optional but if included it needs all three axes, changed to `<val>`

   *    [x], [y], and [z] must be valid numbers and are space separated
   */
  position: PropTypes.string,

  /** Position of the dataset in the scene as a "[x] [y] [z]" string
   *    [x], [y], and [z] must be valid numbers and are space separated
   */
  rotation: PropTypes.string,

  /** Scale of the dataset in the scene as a "[x] [y] [z]" string
   *    [x], [y], and [z] must be valid numbers and are space separated
   */
  scale: PropTypes.string,

  /** Number of slices used to generate the model REQUIRED 

>    *    slices must be a valid integer

Reviewer: must be positive?

Author: Yes! Definitely. Will update the validation function as well

   */
  slices: PropTypes.number.isRequired,

  /** Spacing between the slices of the models a "<x> <y> <z>" string (REQUIRED)
   *    x, y, and z must be valid numbers and are space separated
   */
  spacing: PropTypes.string.isRequired,

  /** 
   * Sliders for control of clipping along the x, y, and z axes 
   * SLIDER is an array of exactly two values between 0 and 1. slider[0] must be <= slider[1].
   *  slider[0]: Minimum slider value
   *  slider[1]: Maximum slider value

>    */

Reviewer: make sure there's an issue open to change this to `clipping`

Author: #95

## Props

The only required props are the model's path, colorMap, and it's minimum and maximum data points. The model's path should be imported into the project and passed in from there - see the [example project](#example).

> CSS styling for the height must be provided and a custom width can be provided as well. You can define your own class and/or pass inline styles into the `<VolumeViewer />` component to accomplish this.

Reviewer: ```suggestion
CSS styling for the height must be provided, otherwise the height of the component will be 0px.
```

I'd leave this short - people can figure out the rest based on their css approach.  One thing that would be worth calling out (if applicable) is if the `VolumeViewer` component plays well with styled components, so one could `styled(VolumeViewer)`

Author: Yes it does! I'll add a sentence about styled components

```jsx

VolumeViewer.propTypes = {
  /**

>    * Blending enum exposed to the user

Reviewer: ```suggestion
   * Blending enum
```

   *  None: Don't apply any blending
   *  Add: Apply additive blending
   */
  blending: PropTypes.oneOf(Object.values(Blending)),

  /** Whether or not the controls can be seen */
  controlsVisible: PropTypes.bool,

  /** Array of models loaded into aframe REQUIRED
   *    colorMap: Object containing the path to the current color image applied to the model.
   *      name: Common name of the color map REQUIRED

>    *      path: Path to the color map source image REQUIRED

Reviewer: I'd wrap all the `REQUIRED` in parens so they visually stand apart from the description

Author: What do you mean by wrap?

Reviewer: like `(REQUIRED)`

   *    colorMaps: Array of possible color maps for the model
   *      colorMap must be present in colorMaps

>    *      Each colorMap must have a unique name in colorMaps

Reviewer: ```suggestion
   *      Each colorMap in colorMaps must have a unique name
```

};
```

## Default Props

> The `Model.Default` object will be merged with every `model` in the `models` array. Note, however, that there is no default value for `colorMap`, `description`, `name`, or `path`. Each of these is required.

Reviewer: it seems to me a default colorMap of grayscale would be reasonable - what do you think?

Author: Planning on tackling more of the default values of `colorMap` and `transferFunction` based on `useColorMap` and `useTransferFunction` in a different PR. I'm inclined to agree though, that's an easy change I'll add in.

Author: Had to change a bit more to get things working properly but it's all set!

```jsx
VolumeViewer.defaultProps = {
  blending: Blending.Add,
  controlsVisible: false,
  position: "0 0 0",
  rotation: "0 0 0",
  scale: "1 1 1",
  sliders: {
    x: [0, 1],
    y: [0, 1],
    z: [0, 1],
  };,
};
```

The default model is merged with every `model` in the `models` array. Note, however, that there is no default value for `colorMap`, `name`, or `path`. Each of these is required.

```js
  {
    colorMaps: [],
    description: "",
    enabled: true,
    intensity: 1,
    range: { min: 0, max: 1, unit: "" },
    transferFunction: [new Point(0, 0), new Point(1, 1)],
    useTransferFunction: true,
    useColorMap: true,
  }
```

## Exports

### VolumeViewer

`<VolumeViewer />` is the main component exported by this library. It's expected props are detailed above.

## Blending

> The `Blending` class is used as an enum for the different algorithms that can be used to blending the models together. `Blending.None` does not apply any blending. `Blending.Add` applies additive blending. Currently only `Blending.None` is implemented.

Reviewer: ```suggestion
The `Blending` object is used as an enum for the different algorithms that can be used to blending the models together. `Blending.None` does not apply any blending. `Blending.Add` applies additive blending. Currently only `Blending.None` is implemented.
```

> ```js
> const Blending = {
>   None: 0,
>   Add: 1,
> };
> ```

Reviewer: ```suggestion
```

I don't think this is needed, keep the implementation details private so you can change them at will 😛

Author: Okay! I'll delete it

### Model

> The `Model` class holds all of the properties for a single model. Certain default values are merged automatically in the constructor and the class comes with it's own validation. Note that you can use `Model.Default` when you are building your own `Model` instances.

Reviewer: I thought we were changing this make to just an object?

```js
class Model {
  static Default = new Model({

>     colorMaps: [],

Reviewer: no need to copy the whole implementation to the readme, just an overview of how to use salient methods/propeties

      obj.transferFunction ?? Model.Default.transferFunction;
    this.useTransferFunction =
      obj.useTransferFunction ?? Model.Default.useTransferFunction;
    this.useColorMap = obj.useColorMap ?? Model.Default.useColorMap;
  }
  toString() {
    return `Model.${this.name}: ${this.path}`;
  }
  validate() {
    // Validate points along the transferFunction
    this.transferFunction.forEach((point) => point.validate());

    if (!this.colorMaps.includes(this.colorMap))
      throw new Error("Color Map '" + this.colorMap + "' not in colorMaps");

    // Color map names in colorMaps must be unique
    const colorMapNames = new Set();
    this.colorMaps.forEach((colorMap) => {
      if (colorMapNames.has(colorMap.name))
        throw new Error(
          "Color map name '" +
            colorMap.name +
            "' is not unique on model '" +
            this.name +
            "'"
        );
      else colorMapNames.add(colorMap.name);
    });
  }
}
```

### COLOR_MAPS

`COLOR_MAPS` is an object containing some example colormaps. Any/all of the color maps can be imported into your project and passed into `model.colorMaps` array for any individual model.

> <img alt="grayscale" src="./src/images/grayscale.png" height="25"/>
> <img alt="natural" src="./src/images/natural.png" height="25"/>
> <img alt="rgb" src="./src/images/rgb.png" height="25"/>

Reviewer: ```suggestion
`Grayscale`
<img alt="grayscale" src="./src/images/grayscale.png" height="25"/>

`Natural`
<img alt="natural" src="./src/images/natural.png" height="25"/>

`Rgb`
<img alt="rgb" src="./src/images/rgb.png" height="25"/>
```

> ```js
> const COLOR_MAPS = {
>   Grayscale: {name: "Grayscale", path: grayscale},
>   Natural: {name: "Natural", path: natural},
>   Rgb: {name: "Rgb", path: rgb},
> }
> ```

Reviewer: ```suggestion
```

The `DEFAULT_SLIDERS` export is the default value for the `sliders` prop. It will be applied automatically if you do not pass `sliders` into `<VolumeViewer />`.

```js
const DEFAULT_SLIDERS = {
  x: [0, 1],
  y: [0, 1],
  z: [0, 1],
};
```

### DEFAULT_MODEL

The `DEFAULT_MODEL` object contains default values for some properties of an individual model. These values are automatically merged with each object in the `models` prop passed in if any specific property is not given. Note that the `colorMap`, `name`, and `path` properties are not present as these are required properties.

```js
const DEFAULT_MODEL = {
  colorMaps: [],
  description: "",
  enabled: true,
  intensity: 1,
  range: { min: 0, max: 1, unit: "" },
  transferFunction: [
    { x: 0, y: 0 },
    { x: 1, y: 1 },
  ],
  useTransferFunction: true,
  useColorMap: true,
}
```

## Example

```jsx
import React from 'react'
import styled from 'styled-components'
import {
  VolumeViewer,
  ColorMap,
  Point,
  DEFAULT_COLOR_MAPS,
  Model,
} from "react-volume-viewer";

import model1 from "./path/to/model.png";
import model2 from "./path/to/model.png";

const haline = new ColorMap("Haline", "./assets/colormaps/haline.png")

> const thermal = new ColorMap("Thermal", "./assets/colormaps/thermal.png")

Reviewer: this looks out of date - make an object?

Author: Very out of date, just fixed everything in the README

function App() {
  const [controlsVisible, setControlsVisible] = React.useState(true);
  const [enabled, setEnabled] = React.useState(true)

  return (
    <StyledVolumeViewer
      controlsVisible={controlsVisible}
      models={[
        new Model({
          name: "Salt",
          colorMap: haline,
          colorMaps={[haline, thermal, ...ColorMap.all]}
          description: "Model visualizing salinity data",
          range: {
            min: 0.05,
            max: 33.71,
          },
          path: model1,
          transferFunction: [new Point(0, 0), new Point(0.5, 0.5), new Point(1, 1)],
        }),
        new Model({
          name: "Temperature",
          colorMap: ColorMap.Grayscale,
          colorMaps={ColorMap.all}
          enabled: enabled,
          description: "Model visualizing temperature data",
          range: {
            min: 2.5,
            max: 42,
            unit: "°C",
          },
          useTransferFunction: false,
          useColorMap: false,
        }),

>       ]}

Reviewer: looks out of date

Author: Updated now

## Blending

The `Blending` object is used as an enum for the different algorithms that can be used to blending the models together. `Blending.None` does not apply any blending. `Blending.Add` applies additive blending. Currently only `Blending.None` is implemented.

### COLOR_MAPS

`COLOR_MAPS` is an object containing some example colormaps. Any/all of the color maps can be imported into your project and passed into `model.colorMaps` array for any individual model.

`Grayscale`
<img alt="grayscale" src="./src/images/grayscale.png" height="25"/>

`Natural`
<img alt="natural" src="./src/images/natural.png" height="25"/>

`Rgb`
<img alt="rgb" src="./src/images/rgb.png" height="25"/>

### DEFAULT_MODEL

The `DEFAULT_MODEL` object holds all of the default properties for a single model.

> It will be merged with every `model` in the `models` array. Note, however, that there is no default value for `colorMap`, `name`, or `path`. Each of these properties are required.

Reviewer: ```suggestion
It will be merged with every `model` in the `models` array. Note, however, that there is no default value for `name`, or `path`. Each of these properties are required.
```

    <StyledVolumeViewer
      controlsVisible={controlsVisible}
      models={[
        {
          name: "Salt",
          colorMap: haline,
          colorMaps={[haline, thermal, COLOR_MAPS.Grayscale]}
          description: "Model visualizing salinity data",
          path: {model1},
          range: {
            min: 0.05,
            max: 33.71,
          },
          path: model1,

>           transferFunction: [new Point(0, 0), new Point(0.5, 0.5), new Point(1, 1)],

Reviewer: the Points should just be x,y objects I think

Author: Good catch!

        },
        {
          name: "Temperature",

>           colorMaps={...Object.values(COLOR_MAPS)}

Reviewer: ```suggestion
          colorMaps={[...Object.values(COLOR_MAPS)]}
```
Is this correct?

Author: Yup! Commited
